### PR TITLE
Update docs for `global-variable-not-assigned`

### LINKED ISSUE #5073: False positive for global-variable-not-assigned when using with
### Bug description

```python
# File a.py
"""
The "global-variable-not-assigned" check does not take into account that a "with"
statement can very well change the global variable.
For example, when using a global condition variable by multiple threads:
"""

import threading

COND = threading.Condition()


def func1():
    "Function that is called in one thread"
    # pylint: disable=global-statement
    global COND
    with COND:
        COND.notify()
```

### Configuration

_No response_

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:16:4: W0602: Using global for 'COND' but no assignment is done (global-variable-not-assigned)
```

### Expected behavior

Pylint realizes that "with" may change the global variable and thus counts it like it was an assignment, and thus does not raise the issue in this case.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.0
Python 3.9.7 (default, Sep  3 2021, 12:45:31) 
[Clang 12.0.0 (clang-1200.0.32.29)]
```

### OS / Environment

macOS

### Additional dependencies

Output of pipdeptree:
pylint==1.6.4
  - astroid [required: >=1.4.5,<1.5.0, installed: 1.4.9]
    - lazy-object-proxy [required: Any, installed: 1.4.2]
    - six [required: Any, installed: 1.15.0]
    - wrapt [required: Any, installed: 1.11.2]
  - backports.functools-lru-cache [required: Any, installed: 1.6.4]
  - configparser [required: Any, installed: 4.0.2]
  - isort [required: >=4.2.5, installed: 4.2.15]
  - mccabe [required: Any, installed: 0.6.1]
  - six [required: Any, installed: 1.15.0]

### PR DESCRIPTION
**Pull request by clavedeluna**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :scroll: Docs          |

## Description

Update explanation for ``global-variable-not-assigned`` and add inference.

Closes #5073

### CODE REVIEW

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -434,8 +434,9 @@ def _has_locals_call_after_node(stmt: nodes.NodeNG, scope: nodes.FunctionDef) ->
     "W0602": (
         "Using global for %r but no assignment is done",
         "global-variable-not-assigned",
-        'Used when a variable is defined through the "global" statement '
-        "but no assignment to this variable is done.",
+        'Used when a name that is previously defined with a "global" statement '
+        "is being invoked with the `global` keyword but is not reassigned in"
+        "the current code block. For example, modifying a global dict.",
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/variables.py:439]:
```suggestion
        'When a variable defined in the global scope is modified in an inner scope' "
        "the 'global' keyword is required in the inner scope only if there is an "
        "assignment operation done in the inner scope.",
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
I have a suggestion for the new explanation, let's wait for other opinion before applying and regenerating the doc :)

File: pylint/checkers/variables.py (hunk 1 above)
Comment (mbyrnepr2) [reply pylint/checkers/variables.py:439]: I like it 👌

File: pylint/checkers/variables.py [hunk 2]
```diff
@@ -434,8 +434,9 @@ def _has_locals_call_after_node(stmt: nodes.NodeNG, scope: nodes.FunctionDef) ->
     "W0602": (
         "Using global for %r but no assignment is done",
         "global-variable-not-assigned",
-        'Used when a variable is defined through the "global" statement '
-        "but no assignment to this variable is done.",
+        "When a variable defined in the global scope is modified in an inner scope"
```
Comment (mbyrnepr2) [inline pylint/checkers/variables.py:437]:
```suggestion
        "When a variable defined in the global scope is modified in an inner scope "
```

**Review by mbyrnepr2 — CHANGES_REQUESTED** [review]
Looks good. Just need to add a space to prevent the two words sticking together 😁

File: pylint/checkers/variables.py (hunk 2 above)
Comment (mbyrnepr2) [reply pylint/checkers/variables.py:437]: Suggesting a comma too because the sentence is a mouthful

File: doc/whatsnew/fragments/5073.other [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Update explanation for ``global-variable-not-assigned`` and add inference.
```
Comment (Pierre-Sassoulas) [inline doc/whatsnew/fragments/5073.other:1]:
```suggestion
Update explanation for ``global-variable-not-assigned`` and add confidence.
```

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you @clavedeluna !

**Pull request merged**